### DISCUSSION TITLE
docker vs self hosted

### DISCUSSION THREAD

**Discussion by ariel256** [netbox-community/netbox#12172] [General]
Hello,
can you please advice if for production use it is better to deploy netbox full server install on OS or docker can be used too?
i.e. are there any limitations or disadvantages to use docker? 
- can i backup and restore netbox database in docker?
- can i install plugins and scripts and reports? 
- can i connect it to LDAP?
- is it more easy or complex to upgrade netbox versions in docker?
- any other considerations or tips you can share?
thanks in advance

**Comment by kkthxbye-code** [comment]
> - can i backup and restore netbox database in docker?

Yes.

> - can i install plugins and scripts and reports? 

Scripts and reports have to be volume mounted to the container. Plugins require that you build a custom image based on the netbox-docker image. It's all in the wiki: https://github.com/netbox-community/netbox-docker/wiki

> - can i connect it to LDAP?

Yes, LDAP auth works just as it does on a native install.

> - is it more easy or complex to upgrade netbox versions in docker?

I would say easier. In most cases it will just be pulling the new image and deploying it. In cases where there are steps to be taken before upgrading, it might require a bit more familiarity with docker commands.

> - any other considerations or tips you can share?

I would recommend running it the way your organization tends to run stuff. If you use docker in other places, it would be fine to run it using docker. Otherwise I would perhaps consider a native install. It all depends on the skills of the system administrators at your organization.

**Comment by cyberndj** [comment]
Howdy,
In my organization, we (aka, I) decided to use docker to do our deploy. Follow the plugins section in the wiki to get it work right. We don't use the latest tag, but instead pin the version down (netboxcommunity/netbox:v3.4.6) in the dockerfile.

We took NetBox a bit further than the instructions by also deploying Portainer, NGINX Proxy Manager (NPM) and pgadmin for further system manageability. We loaded a internally signed cert with SANs on NPM so we can avoid the hassle of installing a cert in Nebox and also use the features of NPM. The two other docker Stacks we have running are Batfish and CheckMK (FOSS version/demo).

We also run 7+ plugins, AD integrated and also use the NAPALM feature. We have a few reports that are occasionally run that we found. (The lack of seeing changes from report to report doesn't make reports useful for us.) PM me if  you want to see working files.

**Reply by jasser-chaoued** [reply]
Hello, i would really like to see how you did all that if it's possible !

**Reply by cyberndj** [reply]
https://github.com/cyberndj/dockers/tree/main/net-management/netbox

I have a folder at /var/lib/docker/netbox (all files go in here unless stated// you can change this folder location to fit your needs)
On the Dockerfile-Plugins file, update the firstline to a specific version or `FROM netboxcommunity/netbox:latest` to use latest.

plugins.py will go into the "nebox-configuration" volume
I use Portainer, so the compose.txt is what is in the stack configuration.

To Run:
1. Update the local-requirements.txt with your desired plugins & edit plugins.py if necessary.
2. Create a custom image: run `docker-compose build --no-cache` from /var/lib/docker/netbox to create your custom image with plugins. 
3. In Portainer, create a new Stack and deploy. (or update the stack if its already created)  It should pull the custom image you made.
4. Do any updates to a reverse proxy if you are using one or update/uncomment out the port settings.

**Comment by m-shalenko** [comment]
Is it possible to deploy netbox on three virtual machines instead of Kubernetes or kind of that using just `docker run` via systemd, for example? Or it requires to have common volumes for netbox media files, reports files and scriptsfiles in main netbox container, rq worker and housekeeping?

**Reply by ziggekatten** [reply]
We have been running Netbox in docker swarm with commmon fileshares (scrips, media, redis data) for years. You really need the possibility for shared filesystems, either through some storage plugin or standard NFS. If you have three nodes an simple swarm cluster of three master nodes would work well.

**Reply by m-shalenko** [reply]
@ziggekatten 

> We have been running Netbox in docker swarm with commmon fileshares (scrips, media, redis data) for years. You really need the possibility for shared filesystems, either through some storage plugin or standard NFS. If you have three nodes an simple swarm cluster of three master nodes would work well. 

Are you sure that it requires to have common shared filesystem? Because I think that every single container that was created from official netbox image has all static and media content inside and in principle, it is enough, isn't it?

**Reply by m-shalenko** [reply]
I think I can use S3 as a storage backend instead of local as well, right?
